datafusion/physical-expr/src/partitioning.rs

```diff
@@ -152,6 +152,8 @@ impl Partitioning {
         match required {
             Distribution::UnspecifiedDistribution => true,
             Distribution::SinglePartition if self.partition_count() == 1 => true,
```

Comment:
`Distribution::SinglePartition if self.partition_count() != 1` is an invalid state. However, during planning we have temporary states where this can happen. Consider the plan
```
GlobalLimit(SingleRequirement)
--FilterExec(a=2)
----Source(n_partitions=1)
```
`EnforceDistribution` rule turns following plan into plan below
```
GlobalLimit(SingleRequirement)
--CoalescePartitions 
----FilterExec(a=2)
------RepartitionExec(n_out=8, n_in=1)
--------Source(n_partitions=1)
```

by generating below intermediate plans
```
GlobalLimit(SingleRequirement)
--FilterExec(a=2)
----RepartitionExec(n_out=8, n_in=1)
------Source(n_partitions=1)
```
```
GlobalLimit(SingleRequirement)
--CoalescePartitions 
----FilterExec(a=2)
------RepartitionExec(n_out=8, n_in=1)
--------Source(n_partitions=1)
```
The first intermediate plan is an invalid plan. However, end plan is valid. In short, we can have invalid states during calls to `fn satisfy`. 
What is important is that after planning we should be in valid state. The [PR](https://github.com/synnada-ai/datafusion-upstream/pull/24) @yfy- is working adds this sanity check to the end of the optimizations stage to verify whether final plan indeed satisfy all invariants.